backend/utils/bulk_operations.py

```
@@ -0,0 +1,295 @@
+"""
+Bulk Operations Utility
+
+This module provides utilities for performing bulk database operations
+to improve performance and avoid N+1 query problems.
+"""
+
+import logging
+from datetime import datetime, timedelta
+from models import db, UserActivity, AuditLog, Tool, Chemical
+from sqlalchemy import and_, or_, func
```

Ruff (0.11.9)

`sqlalchemy.or_` imported but unused

Remove unused import: `sqlalchemy.or_`

(F401)